File: backend/internal/access/permissions.go
package access

// Permission keys used across HTTP handlers and websocket guards.
const (
	PermDeploymentsCreate  = "deployments.create"
	PermDeploymentsDelete  = "deployments.delete"
	PermDeploymentsRestart = "deployments.restart"

	PermDockerManage = "docker.manage"

	// PermPm2Manage covers native host PM2 discovery, process control, logs, and /ws/pm2.
	PermPm2Manage = "pm2.manage"

	PermTerminalAccess = "terminal.access"
	PermMetricsView    = "metrics.view"

	PermFilesystemRead  = "filesystem.read"
	PermFilesystemWrite = "filesystem.write"

	PermUsersManage    = "users.manage"
	PermSettingsManage = "settings.manage"

	PermMarketplaceInstall = "marketplace.install"
	PermServersManage      = "servers.manage"
)

// RoleName constants align with database seed values.
const (
	RoleOwner     = "owner"
	RoleAdmin     = "admin"
	RoleDeveloper = "developer"
	RoleViewer    = "viewer"
)

File: frontend/src/lib/permissions.ts
/** Permission keys issued by the SkyPort API (see internal/access). */
export const PERMS = {
  deploymentsCreate: "deployments.create",
  deploymentsDelete: "deployments.delete",
  deploymentsRestart: "deployments.restart",
  dockerManage: "docker.manage",
  terminalAccess: "terminal.access",
  metricsView: "metrics.view",
  filesystemRead: "filesystem.read",
  filesystemWrite: "filesystem.write",
  usersManage: "users.manage",
  settingsManage: "settings.manage",
  marketplaceInstall: "marketplace.install",
  serversManage: "servers.manage",
} as const;

export type PermissionKey = (typeof PERMS)[keyof typeof PERMS];

export function can(permissions: string[] | undefined | null, key: string): boolean {
  if (!permissions?.length) return false;
  return permissions.includes(key);
}

export function canAny(permissions: string[] | undefined | null, keys: string[]): boolean {
  return keys.some((k) => can(permissions, k));
}
